enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,386 @@
+- Grant the application administrator the ability to tighten cluster-scoped policies. This could be useful in the future, but we are deferring it to limit the amount of work needed for an initial implementation.
+
+## Proposal
+
+### Workflow Description
+
+**cluster administrator** is a human user responsible for adding signature verification configurations in
+cluster scope.
+1. The cluster administrator creates ImagePolicy yaml file with `namespace: openshift-config`.
+2. The cluster administrator creates ImagePolicy instance using `oc create -f imagepolicy.yaml`.
+3. The cluster administrator can get the merged cluster scope policies from `status` of ImagePolicy instances. 
+4. The cluster administrator can delete the signature verification configuration by deleting its ImagePolicy instances. 
+
+**application administrator** is a human user responsible for
+adding signature verification configurations for specific namespace.
+
+1. The application administrator creates ImagePolicy yaml file for namespace that is not `namespace: openshift-config`
+2. The application administrator creates ImagePolicy instance using `oc create -f imagepolicy.yaml`.
+3. The application administrator can get the merged cluster policy and policy for certain namespace from `status` of ImagePolicy instances. 
+4. The application administrator can delete the signature verification configuration by deleting its ImagePolicy instances.
+5. The application administrator cannot override cluster scoped policies because they will treated with a higher priority.
+
+### API Extensions
+
+#### Type definitions
+
+```go
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ImagePolicy holds namespace-wide configuration configuration for image signature verification
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +openshift:enable:FeatureSets=TechPreviewNoUpgrade
+type ImagePolicy struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ImagePolicySpec `json:"spec"`
+	// status contains the observed state of the resource.
+	// +optional
+	Status ImagePolicyStatus `json:"status,omitempty"`
+}
+
+// ImagePolicySpec is the specification of the ImagePolicy CRD.
+type ImagePolicySpec struct {
+	// scopes defines the list of image identities assigned to a policy. Each item refers to an image or repository in a registry implementing the "Docker Registry HTTP API V2".
```

Comment:
@saschagrunert @mtrmac  do we consider supporting transport besides docker or leave space for future implementation? I think that other transports are for images from local storage locations

Reply:
I don't see a use case other than the docker scope. But maybe @mtrmac has something in mind.

Reply:
> Even with only a docker scope supported you can simplify the regex with DockerScope struct. To make sure a namespace is specified when an image is, you can use +kubebuilder:validation:XValidation:rule validation.

@ingvagabund  This can simplify the regex to validate the scopes. Will this make it too complicated for the customers to write the yaml?

Reply:
Good point. Agree customers will prefer to use a simpler way of specifying the string. Feel free to dismiss the `DockerScope`.

Reply:
I don’t think we need to consider non-Docker scopes much. As it happens, I was looking into what CRI-O does, and while it, strictly speaking, does allow using other transports (via a `DefaultTransport` option), the implementation is, for a long time, only accepting a registry-like syntax, and only for transports which can map to a registry reference (because the remote image, using an arbitrary transport, must be pulled into local storage, and local storage names look like registry references). I’m this >< close to proposing to hard-code the `docker://` transport in CRI-O.

I’m tempted to say that if we ever needed to support other transports, that’s unlikely enough that we could deal with that by introducing a whole new top-level CRD, `SomeOtherTransportImagePolicy`. That’s always an option, isn’t it?

---

If anything, I’m worried about adding more wildcard syntaxes, or something like that, not about adding more transports. That did happen before.

But, do we need an enum here at all? Or would it be valid WRT the API evolution rules, instead, to add a new `ImagePolicySpec.SomeNewKindOfScopes` field?

---

> ```go
> type DockerScope struct {
> ```

Just so that this naming doesn‘t propagate, `s/DockerScope/RegistryScope/` in any such things (if we needed a new field), please, to keep one of my colleagues happy.

Reply:
> Or, is it possible one might be allowed/need to provide more configuration in the future?

The scope key in `policy.json` is a single string. _Hypothetically_, that doesn’t prevent that string from carrying multiple fields using some built-in syntax, where OCP might indeed want to use a struct with individual fields.

WRT `ImagePolicy.Spec.Scopes` being an array of strings vs. an array of structs, I don’t have a strong opinion. I think it’s pretty unlikely that we would ever want to add a field, OTOH in a completely different application I did see us painted into a corner because values were strings with no ability to extend, so I have some sympathy for wanting to be careful/extensible here.

_If_ the new top-level field (`SomeNewKindOfScopes` above) is an option, I think wanting to extend scopes is sufficiently unlikely that a a more ergonomic array of strings is preferable; that’s a pretty weak opinion.